misk/src/main/kotlin/misk/web/actions/ReadinessCheckAction.kt

```diff
@@ -37,20 +47,42 @@ class ReadinessCheckAction @Inject internal constructor(
       // Don't do healthchecks if services haven't all started. The app isn't in a good state yet,
       // and a health check could end up triggering random errors that we don't want to flood the
       // logs with.
-      return Response("", statusCode = 503)
+      return@runBlocking Response("", statusCode = 503)
+    }
+
+    // First run needs to block and get statuses
+    if (lastUpdate == null) {
+      refreshStatuses()
+    } else {
+      // This needs to be independent of the max age check so that services can recover
+      // If refreshStatuses() is never run then a failed/stale check can never recover
+      if (clock.instant().isAfter(lastUpdate!!.plusMillis(config.readiness_refresh_interval_ms.toLong()))) {
+        // Launch in the background so that we can still return the cached state
+        launch(Job()) {
```

Comment:
Should we get a thread from misk's health executor thread pool instead?

Reply:
[Added a comment below](https://github.com/cashapp/misk/pull/2883#discussion_r1272733408) about installing the thread pool as an executor service so that it's injectable, then using that instead of the coroutine job